File: C++/题目1251：序列分割.cpp

#include <stdio.h>
#include <string.h>
#include<algorithm>
using namespace std;
const int maxn = 65;
int array[maxn];
int visit[maxn];
int n;
int tempSum, group;

bool compare(int a , int b){
	return a > b;
}

bool dfs(int groupId, int allSum ,int start) {
	if (groupId == group) {
		return true;
	}
	int i , newSum;
	for (i = start; i < n; i++) {
		if (visit[i] != 0) {
			continue;
		}
		if ((newSum = allSum + array[i] ) < tempSum) {
			visit[i] = groupId;
			if (dfs(groupId, newSum ,i+1)) {
				return true;
			}
			visit[i] = 0;	
			if(start == 0){
				break;
			}
		}else if (newSum == tempSum) {
			visit[i] = groupId;
			groupId++;
			if (dfs(groupId, 0  ,0)) {
				return true;
			}
			visit[i] = 0;
			break;
		}
	}
	return false;
}
int main(){
	n = 0;
	while(scanf("%d",&n) != EOF){
		if(n == 0){
			break;
		}
		memset(array ,0 , sizeof(array));
		int maxValue = 0;
		int sum = 0;
		for(int i = 0 ; i < n; i++){
			scanf("%d",&array[i]);
			sum += array[i];
		}
		sort(array, array + n, compare);
		int resultM = 1;
		for (int j = array[0]; j <= sum ; j++) {
            if (sum % j == 0) {
				tempSum = j;
                memset(visit ,0 , sizeof(visit));
				group = sum / j;
				bool flag = dfs(1, 0 ,0);
                if (flag) {
					resultM = group;
                    break;
                }
            }
        }
		printf("%d\n",resultM);
	
	}
	return 0;
}
/**************************************************************
	Problem: 1251
	User: wzqwsrf
	Language: C++
	Result: Accepted
	Time:10 ms
	Memory:1020 kb
****************************************************************/


                        

File: Java/题目1251：序列分割.java

// 题目1251：序列分割

/**
 * @author:wangzq
 * @email:wangzhenqing1008@163.com
 * @date:2015-06-30 11:01:54
 * @url:http://ac.jobdu.com/problem.php?pid=1251
 */

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;
import java.util.Arrays;
import java.util.Collections;

public class Main {
    /*
     * 1251
     */
    private static Integer array[];

    private static int visit[];

    private static int n;

    private static int tempSum;

    private static int group;

    public static void main(String[] args) throws Exception {
        StreamTokenizer st = new StreamTokenizer(new BufferedReader(
                new InputStreamReader(System.in)));
        while (st.nextToken() != StreamTokenizer.TT_EOF) {
            n = (int) st.nval;
            if (n == 0) {
                break;
            }
            int sum = 0;
            array = new Integer[n];
            for (int i = 0; i < n; i++) {
                st.nextToken();
                array[i] = (int) st.nval;
                sum += array[i];
            }
            Arrays.sort(array, Collections.reverseOrder());
            int resultM = 1;
            for (int i = array[0]; i <= sum; i++) {
                if (sum % i == 0) {
                    tempSum = i;
                    group = sum / i;
                    visit = new int[n];
                    if (dfs(1, 0, 0)) {
                        resultM = group;
                        break;
                    }
                }
            }
            System.out.println(resultM);
        }
    }

    private static boolean dfs(int groupId, int sum, int start) {
        if (groupId == group) {
            return true;
        }
        for (int i = start; i < n; i++) {
            if (visit[i] != 0) {
                continue;
            }
            int needSum;
            if ((needSum = sum + array[i]) < tempSum) {
                visit[i] = groupId;
                if (dfs(groupId, needSum, i + 1)) {
                    return true;
                }
                visit[i] = 0;
                if (start == 0) {
                    break;
                }
            } else if (needSum == tempSum) {
                visit[i] = groupId;
                groupId++;
                if (dfs(groupId, 0, 0)) {
                    return true;
                }
                visit[i] = 0;
                break;
            }
        }
        return false;
    }
}

/**************************************************************
 Problem: 1251
 User: wzqwsrf
 Language: Java
 Result: Accepted
 Time:120 ms
 Memory:20104 kb
 ****************************************************************/




